Feature/checkin form endpoints

curtwl opened the pull request:
# Description

Retrieves check-in form responses given a sprintNumber, voyageId, userId, or teamId query param or any combination of them.
I also added a global function that does the basic if !itemFound throw 404 logic given a table name and key/value. 

## Issue link

https://app.clickup.com/t/86azuu2yc

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Feature updates / changes
- [ ] Tests
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update


# How Has This Been Tested?

New tests and manual verification with Swagger that errors are handled as expected.


# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have updated the change log

Ajen07 commented on src/sprints/sprints.response.ts:
```diff
@@ -222,3 +222,32 @@ export class CheckinSubmissionResponse {
     @ApiProperty({ example: "2023-11-30T06:47:11.694Z" })
     createdAt: Date;
 }
+
+export class CheckinFormResponse {
+    @ApiProperty({ example: 1 })
+    id: number;
+
+    @ApiProperty({ example: 1 })
+    voyageTeamMemberId: number;
+
+    @ApiProperty({ example: 1 })
+    voyageTeamId: number;
+
+    @ApiProperty({ example: 1 })
+    sprintId: number;
+
+    @ApiProperty({ example: "2023-11-30T06:47:11.694Z" })
+    adminComments: string;
+
+    @ApiProperty({ example: "2023-11-30T06:47:11.694Z" })
+    feedbackSent: boolean;
+
+    @ApiProperty({ example: "2023-11-30T06:47:11.694Z" })
+    responseGroupId: number;
+
+    @ApiProperty({ example: "2023-11-30T06:47:11.694Z" })
+    createdAt: Date;
+
+    @ApiProperty({ example: "2023-11-30T06:47:11.694Z" })
+    updatedAt: Date;
+}
```
adminComments , feedbackSent , responseGroupId  examples can be updated to their respective types rather having a date-time format for each one

JoshuaHinman reviewed (commented):
When I ran unit tests, these lines in `sprints.contoller` threw errors:
![image](https://github.com/chingu-x/chingu-dashboard-be/assets/26824124/dcabaab8-55dd-4340-9466-2340d9cdc46f)
I changed the paths form `src/` to `../` to get them working

curtwl commented:
> When I ran unit tests, these lines in `sprints.contoller` threw errors: ![image](https://private-user-images.githubusercontent.com/26824124/341209866-dcabaab8-55dd-4340-9466-2340d9cdc46f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkwMjE3NzksIm5iZiI6MTcxOTAyMTQ3OSwicGF0aCI6Ii8yNjgyNDEyNC8zNDEyMDk4NjYtZGNhYmFhYjgtNTVkZC00MzQwLTk0NjYtMjM0MGQ5Y2RjNDZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIyVDAxNTc1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxNDFkZjhiMzFlOGM2YjU1MmU1ZTBjZTFkMDU1MzA1ZDJjNDZiZmFkZjIzMmMwMGU3MDVhNjIxZWU1ZWZkYmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KjT0EE_Ukcv5b51HoOwPtn-KbAoZAVshKjTjiH4WZl8) I changed the paths form `src/` to `../` to get them working

Strange, both ways work for me... I've updated them to use ../ instead of src/

Ajen07 reviewed (approved):
Great Job ! All the changes looks good.

cherylli reviewed (changes requested):
accidentally pressed the approved button

curtwl commented:
> 1. Is it possible also populate the responses? In the response I can only see `responseGroupId:1`,  this doesn't really allow the admin to see the responses when they pull the submission.
> 
> ===========
> 
> 2. I also see something weird, I get the same result regardless of what sprintNumber I put, I always get sprintId: 1 (there's only 1 check in form in my database for sprint 1)
> 
> ![image](https://private-user-images.githubusercontent.com/6191116/342066927-6eafeb35-09c3-4ffe-9aec-b0c10bcdccaf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyNzY0OTMsIm5iZiI6MTcxOTI3NjE5MywicGF0aCI6Ii82MTkxMTE2LzM0MjA2NjkyNy02ZWFmZWIzNS0wOWMzLTRmZmUtOWFlYy1iMGMxMGJjZGNjYWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjVUMDA0MzEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGVmNmZiNTZkMzVkYmQyNjdiM2MxYjJlNTU3NmQ2MjU2MjVmODIzYWEyZTAzMTNhOTdmZmM3NjgwODM4YzVjZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.KzsIzoqWRQJgsfdXZosGaSJ5vrrvG7DrfieV8dOO_7w)
> 
> =============
> 
> 3. when `voyageNumber` is not given, sprint Number should not work, as we don't know sprint x of what voyage, I'm getting records for sprintId 2. Please note that sprint number and sprint id are different.

I see what you mean, will get that fixed up.

JoshuaHinman reviewed (approved):
All tests passed, and I was able to query check-ins using each param - nice work!
I have a quick question about one line from `sprints.service.ts` that I was confused about:
![image](https://github.com/chingu-x/chingu-dashboard-be/assets/26824124/68a21cd7-d967-47f7-8937-c7dbfa7dda22)
 `queryObjectStringified` is an array of strings, right?  I was wondering why/how it extracts key and value from a string

curtwl commented:
> All tests passed, and I was able to query check-ins using each param - nice work! I have a quick question about one line from `sprints.service.ts` that I was confused about: ![image](https://private-user-images.githubusercontent.com/26824124/343980699-68a21cd7-d967-47f7-8937-c7dbfa7dda22.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk4Mjc0MzUsIm5iZiI6MTcxOTgyNzEzNSwicGF0aCI6Ii8yNjgyNDEyNC8zNDM5ODA2OTktNjhhMjFjZDctZDk2Ny00N2Y3LTg5MzctYzdkYmZhN2RkYTIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAxVDA5NDUzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2MzY2ZmY3YWNiNjIzOGI1OGU3OWVlYWNhYjc2NzMxOWM4YTRlNzJiMDI5ZGU1MDBhZTFiZDQzYWE1ZjM0NWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Glt1K9X2nDtqTsDM6Slvk6iuqBe8OmzPgrPxAxUVnts) `queryObjectStringified` is an array of strings, right? I was wondering why/how it extracts key and value from a string

I realized my description in the comment is actually not quite right. JS internally stores the key names as strings already, and this function just puts them (and their values) into an array as they are just to make it easier to do the switch statement. I couldn't think of any way other than that or a really long ternary chain, but if there's a cleaner way to do it, I'd love to know!

cherylli reviewed (changes requested):
I hope this won't hard to fix but this combination returns the wrong thing
![image](https://github.com/chingu-x/chingu-dashboard-be/assets/6191116/377dffe9-042e-4eb0-9c09-b386fbfacacf)


It returns checkin form submission that doesn't belong to `teamId=15`, it seems to ignore teamId,

I created another check in form teamId=8 and both were returned regardless of what teamId I entered

and this for example - 
![image](https://github.com/chingu-x/chingu-dashboard-be/assets/6191116/0f57b3e3-ab9a-4a35-9941-d35bf26cf7b6)

I think it should not return anything, or an error saying user is not in teamId 8, since the teamId is not 8. If there's any contraction e.g. if team 8 is not in v46 we should also return that error

===================================
I also think if the query params are correct but there's no check in form found we should use status 200 with an empty array

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404

> The HTTP 404 Not Found response status code indicates that the server cannot find the requested resource. Links that lead to a 404 page are often called broken or dead links and can be subject to [link rot](https://en.wikipedia.org/wiki/Link_rot).

It would be for cases like teamId does not exist then url would be invalid or broken

curtwl commented:
> I hope this won't hard to fix but this combination returns the wrong thing ![image](https://private-user-images.githubusercontent.com/6191116/346315814-377dffe9-042e-4eb0-9c09-b386fbfacacf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAzMzc2ODgsIm5iZiI6MTcyMDMzNzM4OCwicGF0aCI6Ii82MTkxMTE2LzM0NjMxNTgxNC0zNzdkZmZlOS0wNDJlLTRlYjAtOWMwOS1iMzg2ZmJmYWNhY2YucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDdUMDcyOTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTVkMTAzOWQyMmQ0NWUxNmU4ZGQ4YWIyMzRmYTJlMzU1YzkxNTY3YjUxMzg5NjcwYTNhMmYzODM0NjkzNDUyYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ZvpUla0Ep4C3g0urs-Xt1N76XytesUSVCcbmnlvwaDo)
> 
> It returns checkin form submission that doesn't belong to `teamId=15`, it seems to ignore teamId,
> 
> I created another check in form teamId=8 and both were returned regardless of what teamId I entered
> 
> and this for example - ![image](https://private-user-images.githubusercontent.com/6191116/346315290-0f57b3e3-ab9a-4a35-9941-d35bf26cf7b6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAzMzc2ODgsIm5iZiI6MTcyMDMzNzM4OCwicGF0aCI6Ii82MTkxMTE2LzM0NjMxNTI5MC0wZjU3YjNlMy1hYjlhLTRhMzUtOTk0MS1kMzViZjI2Y2Y3YjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDdUMDcyOTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDUzNmIzNGNhNWVhNzgyYmYxM2FlMDY5MjI5MGFkM2Y0ZmRiZTU4MzY1YjEyZjdiMmI0YzZiNzBkOTU5NzlkMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.-Raku99WWSh2Rv513CLThkwiQEHF5v0qoQBDGHvdQJc)
> 
> I think it should not return anything, or an error saying user is not in teamId 8, since the teamId is not 8. If there's any contraction e.g. if team 8 is not in v46 we should also return that error
> 
> =================================== I also think if the query params are correct but there's no check in form found we should use status 200 with an empty array
> 
> https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404
> 
> > The HTTP 404 Not Found response status code indicates that the server cannot find the requested resource. Links that lead to a 404 page are often called broken or dead links and can be subject to [link rot](https://en.wikipedia.org/wiki/Link_rot).
> 
> It would be for cases like teamId does not exist then url would be invalid or broken
> 
> For example, voyage 46, sprint 6, this resource definitely exists but just no checkin form submissions

I thought from the clickup description (with phrasing like "get all for a particular voyage (all sprints)") that the route was meant to take one of four options (including sprintNumber+voyageNumber as one separately), not combinations of them. So that was by design -- only the first key is used and others are ignored, rather than making the user try again.

Changing 404 to 200 for no check-in forms found but valid params should be easy to fix with some of the global functions. Wasn't sure how to handle that. So it should just return an empty array and frontend will take it from there?

cherylli commented:
> I thought from the clickup description (with phrasing like "get all for a particular voyage (all sprints)") that the route was meant to take one of four options (including sprintNumber+voyageNumber as one separately), not combinations of them. So that was by design -- only the first key is used and others are ignored, rather than making the user try again.

user + voyage number + sprint number should be a valid choice. voyage number + sprint number + team id should also be a valid choice. Sorry I forgot to list this cases. Either way, I don't think it's a good idea to just "ignore" user inputs, we should return an error message saying the combination is not allowed, instead of returning some unexpected results. 

I'm not sure what's the best way to handle it, as a user it feels like i'd be able to search/filter by a combination when the options are presented this way. I guess we sort of deviated a bit when we decide to use the query params.

I think what you're doing now is great as it can handle both cases (get all... and get more specific), but just some common combinations don't seem to be working. 

> So it should just return an empty array and frontend will take it from there? 
yes they could handle cases for array length === 0

cherylli commented on src/sprints/sprints.response.ts:
```diff
@@ -222,3 +223,53 @@ export class CheckinSubmissionResponse {
+    voyage: VoyageNumber;
+}
+
+export class CheckinFormResponse {
```
I think this is not used, so the response is still showing the old ones, 
![image](https://github.com/user-attachments/assets/6ca4cd2d-ea5f-4141-9720-f43987444a1f)
I assume it's supposed to go here
![image](https://github.com/user-attachments/assets/7bb6d326-2143-4188-a5f1-b7f522471d0b)

curtwl replied:
Oh lol, I managed to confuse myself by making the name similar to CheckinSubmissionResponse

cherylli reviewed (changes requested):
I like this approach, we could reuse the logic for other endpoints like fetching solo projects and other forms in future 👍
checked a few random cases and they were fine, all tests passed
just an unused response posted in comment

JoshuaHinman reviewed (changes requested):
I got correct responses with all params, including some combinations I tried.  However, there was one issue I stumbled on: when querying with sprint number 0, it returns all sprints.

curtwl commented:
> I got correct responses with all params, including some combinations I tried. However, there was one issue I stumbled on: when querying with sprint number 0, it returns all sprints.

Fixed, now sprint numbers are restricted to 1-6

cherylli commented:
> > I got correct responses with all params, including some combinations I tried. However, there was one issue I stumbled on: when querying with sprint number 0, it returns all sprints.
> 
> Fixed, now sprint numbers are restricted to 1-6

that reminds me we need to fix create-checkin-form.dto.ts

cherylli reviewed (approved):
Thanks for the quick fix!

curtwl commented:
> > > I got correct responses with all params, including some combinations I tried. However, there was one issue I stumbled on: when querying with sprint number 0, it returns all sprints.
> > 
> > 
> > Fixed, now sprint numbers are restricted to 1-6
> 
> that reminds me we need to fix create-checkin-form.dto.ts
> 
> edit: should probably have asked before pushing my changes but I thought it might be easier if I change it quickly, and it's not one of the files you changed. 🤦‍♀️🙇‍♀️

I tried that locally a few days ago, but I noticed a complication, that someone on i.e. a v48 team can submit check-ins for any sprint id, and end up with check-ins in multiple voyages with that team.

cherylli commented:
> > > > I got correct responses with all params, including some combinations I tried. However, there was one issue I stumbled on: when querying with sprint number 0, it returns all sprints.
> > > 
> > > 
> > > Fixed, now sprint numbers are restricted to 1-6
> > 
> > 
> > that reminds me we need to fix create-checkin-form.dto.ts
> > edit: should probably have asked before pushing my changes but I thought it might be easier if I change it quickly, and it's not one of the files you changed. 🤦‍♀️🙇‍♀️
> 
> I tried that locally a few days ago, but I noticed a complication, that someone on i.e. a v48 team can submit check-ins for any sprint id, and end up with check-ins in multiple voyages with that team.

We probably need to add a check for that, currently we have check for the teamMemberId but not sprintId, we will probably need to check teamMemberId->voyageTeam->voyage->sprintId, but that's a lot more complicated than just restricting the DTO, so this can be its own PR (unless its actually simpler than I thought... 🤔)

JoshuaHinman reviewed (approved):
Retested - everything passed.  Excellent work!
